src/datumaro/plugins/data_formats/datumaro/base.py

```diff
@@ -326,3 +286,173 @@ def _load_annotations(item):
+    ALLOWED_VERSIONS = {LEGACY_VERSION, CURRENT_DATUMARO_FORMAT_VERSION}
+
+    def __init__(
+        self,
+        path: str,
+        *,
+        subset: Optional[str] = None,
+        stream: bool = True,
+        ctx: Optional[ImportContext] = None,
+    ):
+        assert osp.isfile(path), path
+        subset = osp.splitext(osp.basename(path))[0] if subset is None else subset
+        self._stream = stream
+        super().__init__(subset=subset, ctx=ctx)
+        self._init_path(path)
+
+        dm_version = self._get_dm_format_version(path)
+
+        # when backward compatibility happen, we should implement version specific readers
+        if dm_version not in self.ALLOWED_VERSIONS:
+            raise DatasetImportError(
+                f"Datumaro format version of the given dataset is {dm_version}, "
+                f"but not supported by this Datumaro version: {__version__}. "
+                f"The allowed datumaro format versions are {self.ALLOWED_VERSIONS}. "
+                "Please install the latest Datumaro."
+            )
+
+        self._load_impl(path)
+
+    def _init_path(self, path: str):
+        """Initialize path variables"""
+        rootpath = ""
+        if path.endswith(osp.join(DatumaroPath.ANNOTATIONS_DIR, osp.basename(path))):
+            rootpath = path.rsplit(DatumaroPath.ANNOTATIONS_DIR, maxsplit=1)[0]
+        self._rootpath = rootpath
+
+        images_dir = ""
+        if rootpath and osp.isdir(osp.join(rootpath, DatumaroPath.IMAGES_DIR)):
+            images_dir = osp.join(rootpath, DatumaroPath.IMAGES_DIR)
+        self._images_dir = images_dir
+
+        pcd_dir = ""
+        if rootpath and osp.isdir(osp.join(rootpath, DatumaroPath.PCD_DIR)):
+            pcd_dir = osp.join(rootpath, DatumaroPath.PCD_DIR)
+        self._pcd_dir = pcd_dir
+
+    @property
+    def is_stream(self) -> bool:
+        return self._stream
+
+    def infos(self):
+        return self._reader.infos
+
+    def categories(self):
+        return self._reader.categories
+
+    def media_type(self):
+        return self._reader.media_type
+
+    def _load_impl(self, path: str) -> None:
+        """Actual implementation of loading Datumaro format."""
+        self._reader = (
+            DefaultReader(path, self._subset, self._rootpath, self._images_dir, self._pcd_dir)
+            if not self._stream
+            else StreamDefaultReader(
+                path, self._subset, self._rootpath, self._images_dir, self._pcd_dir
+            )
+        )
+        return self._reader
+
+    def _get_dm_format_version(self, path) -> str:
+        """
+        Get Datumaro format at exporting the dataset
+
+        Note that the legacy Datumaro doesn't store the version into exported dataset.
+        Thus it returns DatumaroBase.LEGACY_VERSION
+        """
+        with open(path, "r", encoding="utf-8") as fp:
+            version = json_stream.load(fp).get("dm_format_version", self.LEGACY_VERSION)
```

Comment:
Good to introduce `legacy` version.

Reply:
It produces the default value `self.LEGACY_VERSION` if `dm_format_version` is not specified in the file.